ADC Driver build is broken for STM32L412

**Describe the bug**
Trying to build the ADC driver for a system with an STM32L412 results in an error.

**To Reproduce**
Steps to reproduce the behavior:
1. Use ADC driver in a project based upon a STM32L412
2. Build the project

**Expected behavior**
ADC3 should not be used.

**Impact**
Breaks the build for projects using the ADC driver and a STM32L412

**Logs and console output**

    zephyr/drivers/adc/adc_stm32.c: In function 'adc_stm32_setup_channels':
    zephyr/drivers/adc/adc_stm32.c:858:29: error: 'ADC3' undeclared (first use in this function); did you mean 'ADC1'?
      858 |  if ((adc == ADC1 || adc == ADC3) &&
          |                             ^~~~
          |                             ADC1

**Environment (please complete the following information):**
Zephyr revision df443f1d29e49f69708da1dfb3018ae8d1ec3f30

--- a/drivers/adc/adc_stm32.c
+++ b/drivers/adc/adc_stm32.c
@@ -256,6 +256,8 @@ struct adc_stm32_cfg {
 	void (*irq_cfg_func)(void);
 	struct stm32_pclken pclken;
 	const struct pinctrl_dev_config *pcfg;
+	bool has_temp_channel;
+	bool has_vref_channel;
 };
 
 #ifdef CONFIG_ADC_STM32_SHARED_IRQS
@@ -798,96 +800,17 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
 
 static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_id)
 {
-#if defined(CONFIG_SOC_SERIES_STM32G4X) || \
-	defined(CONFIG_SOC_SERIES_STM32F3X) || \
-	defined(CONFIG_SOC_SERIES_STM32F3X) || \
-	defined(CONFIG_SOC_SERIES_STM32F7X) || \
-	defined(CONFIG_SOC_SERIES_STM32H7X) || \
-	defined(CONFIG_SOC_SERIES_STM32L4X) || \
-	defined(CONFIG_SOC_SERIES_STM32L5X) || \
-	defined(CONFIG_SOC_SERIES_STM32MP1X)
 	const struct adc_stm32_cfg *config = dev->config;
-	ADC_TypeDef *adc = config->base;
-#endif
 
-#if defined(CONFIG_SOC_SERIES_STM32G4X)
-	/* In the G4 model, once can access internal temperature on ADCs 1 and 5. */
-	if ((adc == ADC1 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id) ||
-		(adc == ADC5 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC5) == channel_id)) {
+	if (config->has_temp_channel &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
 		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
 	}
-	/* In the G4 model, once can access internal reference voltage on all ADCs besides 2. */
-	else if (adc != ADC2 &&
+
+	if (config->has_vref_channel &&
 		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
 		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
 	}
-#elif defined(CONFIG_SOC_SERIES_STM32F3X)
-	/* Some F3X have ADC 1 and 2, some have 1, 2, 3 and 4.
-	 * Temperature can only be read on ADC1.
-	 * Refint can be read on all ADCs.
-	 */
-	if (adc == ADC1 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
-		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-	} else if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
-		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
-	}
-#elif defined(CONFIG_SOC_SERIES_STM32F7X)
-	if (adc == ADC1) {
-		if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-		} else if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
-		}
-	}
-#elif defined(CONFIG_SOC_SERIES_STM32H7X)
-	if (adc == ADC3) {
-		if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-		} else if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
-		}
-	}
-
-#elif defined(CONFIG_SOC_SERIES_STM32L4X)
-	/* Temperature available on ADCs 1 and 3.
-	 * Refint available on ADC 1 only.
-	 */
-	if ((adc == ADC1 || adc == ADC3) &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
-		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-	} else if (adc == ADC1 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
-		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
-	}
-#elif defined(CONFIG_SOC_SERIES_STM32L5X)
-	if (adc == ADC1) {
-		if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-		} else if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
-		}
-	}
-#elif defined(CONFIG_SOC_SERIES_STM32MP1X)
-	if (adc == ADC2) {
-		if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-		} else if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
-			adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
-		}
-	}
-#else
-	/* For everything else, temperature and refint are available on all ADCs.
-	 * Most other models only have 1 ADC anyway.
-	 */
-	if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
-		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
-	} else if (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
-		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
-	}
-#endif
 }
 
 static int adc_stm32_channel_setup(const struct device *dev,
@@ -1202,6 +1125,8 @@ static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
 		.bus = DT_INST_CLOCKS_CELL(index, bus),			\
 	},								\
 	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),			\
+	.has_temp_channel = DT_INST_PROP(index, has_temp_channel),		\
+	.has_vref_channel = DT_INST_PROP(index, has_vref_channel),		\
 };
 #else
 #define ADC_STM32_CONFIG(index)						\
@@ -1221,6 +1146,8 @@ static const struct adc_stm32_cfg adc_stm32_cfg_##index = {		\
 		.bus = DT_INST_CLOCKS_CELL(index, bus),			\
 	},								\
 	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),			\
+	.has_temp_channel = DT_INST_PROP(index, has_temp_channel),		\
+	.has_vref_channel = DT_INST_PROP(index, has_vref_channel),		\
 };
 #endif /* CONFIG_ADC_STM32_SHARED_IRQS */
 

--- a/dts/arm/st/f0/stm32f0.dtsi
+++ b/dts/arm/st/f0/stm32f0.dtsi
@@ -371,6 +371,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dma1: dma@40020000 {

--- a/dts/arm/st/f1/stm32f1.dtsi
+++ b/dts/arm/st/f1/stm32f1.dtsi
@@ -314,6 +314,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dma1: dma@40020000 {

--- a/dts/arm/st/f2/stm32f2.dtsi
+++ b/dts/arm/st/f2/stm32f2.dtsi
@@ -363,6 +363,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dma1: dma@40026000 {

--- a/dts/arm/st/f3/stm32f302.dtsi
+++ b/dts/arm/st/f3/stm32f302.dtsi
@@ -87,6 +87,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 	};
 };

--- a/dts/arm/st/f4/stm32f4.dtsi
+++ b/dts/arm/st/f4/stm32f4.dtsi
@@ -490,6 +490,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dma1: dma@40026000 {

--- a/dts/arm/st/f7/stm32f7.dtsi
+++ b/dts/arm/st/f7/stm32f7.dtsi
@@ -688,6 +688,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dac1: dac@40007400 {

--- a/dts/arm/st/g0/stm32g0.dtsi
+++ b/dts/arm/st/g0/stm32g0.dtsi
@@ -359,6 +359,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dma1: dma@40020000 {

--- a/dts/arm/st/g4/stm32g4.dtsi
+++ b/dts/arm/st/g4/stm32g4.dtsi
@@ -81,6 +81,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		adc2: adc@50000100 {

--- a/dts/arm/st/h7/stm32h7.dtsi
+++ b/dts/arm/st/h7/stm32h7.dtsi
@@ -759,6 +759,8 @@
 			status = "disabled";
 			label = "ADC_3";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dac1: dac@40007400 {

--- a/dts/arm/st/l0/stm32l0.dtsi
+++ b/dts/arm/st/l0/stm32l0.dtsi
@@ -283,6 +283,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dma1: dma@40020000 {

--- a/dts/arm/st/l1/stm32l1.dtsi
+++ b/dts/arm/st/l1/stm32l1.dtsi
@@ -214,6 +214,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dac1: dac@40007400 {

--- a/dts/arm/st/l4/stm32l412.dtsi
+++ b/dts/arm/st/l4/stm32l412.dtsi
@@ -32,6 +32,11 @@
 			status = "disabled";
 			label = "SPI_2";
 		};
+
+		adc1: adc@50040000 {
+			has-temp-channel;
+			has-vref-channel;
+		};
 	};
 
 	usb_fs_phy: usbphy {

--- a/dts/arm/st/l4/stm32l432.dtsi
+++ b/dts/arm/st/l4/stm32l432.dtsi
@@ -64,6 +64,11 @@
 			label = "DAC_1";
 			#io-channel-cells = <1>;
 		};
+
+		adc1: adc@50040000 {
+			has-temp-channel;
+			has-vref-channel;
+		};
 	};
 
 	usb_fs_phy: usbphy {

--- a/dts/arm/st/l4/stm32l452.dtsi
+++ b/dts/arm/st/l4/stm32l452.dtsi
@@ -146,6 +146,11 @@
 			sjw = <1>;
 			sample-point = <875>;
 		};
+
+		adc1: adc@50040000 {
+			has-temp-channel;
+			has-vref-channel;
+		};
 	};
 
 	usb_fs_phy: usbphy {

--- a/dts/arm/st/l4/stm32l471.dtsi
+++ b/dts/arm/st/l4/stm32l471.dtsi
@@ -240,6 +240,11 @@
 			#io-channel-cells = <1>;
 		};
 
+		adc1: adc@50040000 {
+			has-temp-channel;
+			has-vref-channel;
+		};
+
 		adc3: adc@50040200 {
 			compatible = "st,stm32-adc";
 			reg = <0x50040200 0x100>;
@@ -248,6 +253,7 @@
 			status = "disabled";
 			label = "ADC_3";
 			#io-channel-cells = <1>;
+			has-temp-channel;
 		};
 	};
 };

--- a/dts/arm/st/l4/stm32l4r5.dtsi
+++ b/dts/arm/st/l4/stm32l4r5.dtsi
@@ -304,6 +304,11 @@
 			label = "DAC_1";
 			#io-channel-cells = <1>;
 		};
+
+		adc1: adc@50040000 {
+			has-temp-channel;
+			has-vref-channel;
+		};
 	};
 
 	otgfs_phy: otgfs_phy {

--- a/dts/arm/st/l5/stm32l5.dtsi
+++ b/dts/arm/st/l5/stm32l5.dtsi
@@ -569,6 +569,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		adc2: adc@42028100 {

--- a/dts/arm/st/u5/stm32u5.dtsi
+++ b/dts/arm/st/u5/stm32u5.dtsi
@@ -368,6 +368,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		adc4: adc@46021000 {
@@ -378,6 +380,8 @@
 			status = "disabled";
 			label = "ADC_4";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		ucpd1: ucpd@4000dc00 {

--- a/dts/arm/st/wb/stm32wb.dtsi
+++ b/dts/arm/st/wb/stm32wb.dtsi
@@ -357,6 +357,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		iwdg: watchdog@40003000 {

--- a/dts/arm/st/wl/stm32wl.dtsi
+++ b/dts/arm/st/wl/stm32wl.dtsi
@@ -310,6 +310,8 @@
 			status = "disabled";
 			label = "ADC_1";
 			#io-channel-cells = <1>;
+			has-temp-channel;
+			has-vref-channel;
 		};
 
 		dac1: dac@40007400 {

--- a/dts/bindings/adc/st,stm32-adc.yaml
+++ b/dts/bindings/adc/st,stm32-adc.yaml
@@ -27,5 +27,13 @@ properties:
     pinctrl-names:
       required: true
 
+    has-temp-channel:
+      type: boolean
+      description: Indicates if the ADC has a dedicated internal temperature sensor channel.
+
+    has-vref-channel:
+      type: boolean
+      description: Indicates if the ADC has a dedicated internal voltage reference channel.
+
 io-channel-cells:
     - input

